### ISSUE TITLE
Memory consumption of initial pair search

### ISSUE THREAD

**Issue by nmoehrle** [simonfuhrmann/mve#264]
The least squares fundamental estimation within the initial pair search becomes a memory bottleneck for scenes that have a large amount of pairwise matches. The cause is the size of the temporary matrices inside the SVD implementation.

For a scene with high resolution images and around 25000 pairwise matches this can lead to a memory consumption of whopping 33 GB:
```
->97.42% (35,549,996,368B) 0x447CB0: void math::internal::matrix_r_svd<double>(double const*, int, int, double*, double*, double*, double const&) (new_allocator.h:104)
| ->97.42% (35,549,996,368B) 0x441EBB: sfm::fundamental_least_squares(std::vector<sfm::Correspondence2D2D, std::allocator<sfm::Correspondence2D2D> > const&, math::Matrix<double, 3, 3>*) (matrix_svd.h:730)
|   ->97.42% (35,549,996,368B) 0x43448C: sfm::bundler::InitialPair::compute_pair(sfm::bundler::InitialPair::Result*) [clone ._omp_fn.0] (bundler_init_pair.cc:226)
```

**Comment by simonfuhrmann** [comment]
It is indeed a problem. @flanggut, any ideas how to solve this?

 * We can limit the size of points we consider to a few hundreds. This should be enough.
 * Maybe we don't need this step and we rely on BA to fix the inaccuracies from not computing the consensus Fundamental matrix.

**Comment by flanggut** [comment]
I think ideally the re-estimation from all matches ensures that we can triangulate as many matches as possible for the initial pair. This hasn't worked out super effective so far - I still see a lot of the successful initial pair searches run into problems in the initial triangulation step (especially in the context of CR-PLAY with not very ideal datasets). I actually implemented a fix to make the search more robust for the context of the project f1db54cd4f03f8bbccf9d90091716267cd5ae673. It simply checks if the triangulation for the initial pair actually works for a significant amount of the matches before it returns the pair.
Overall I'd suggest we do limit input of the estimation to a certain amount of (maybe randomly selected) points, but also check - similar to f1db54cd4f03f8bbccf9d90091716267cd5ae673 - if the triangulation still works for most of the points before we return the pair.

**Comment by nmoehrle** [comment]
#265 limits the amount to 1000 random matches.

Checking the triangulation (f1db54c) solved an initial pair selection issue for me today. Most of the tracks from the automatically selected initial pair have been rejected by the subsequent triangulation which resulted in immediate termination.
```
Using views 305 and 307 as initial pair.
Triangulated 5 new tracks, rejected 26528 bad tracks.
 Rejected 32 tracks with large error.
 Rejected 26494 tracks behind cameras.
 Rejected 2 tracks with unstable angle.
Deleted 0 of 5 tracks (0.00%) above a threshold of 0.0149317.
Running full bundle adjustment...
PBA: MSE 1.33267e-05 -> 4.66928e-06 (), 49 iter, 131ms.
Camera 305, focal length: 0.825675 -> 0.825675, distortion: 0
Camera 307, focal length: 0.825675 -> 0.825675, distortion: 0
SfM reconstruction finished.
SfM reconstruction took 12188 ms.
```

**Comment by flanggut** [comment]
Tested #265 in combination with #270 without problems on several datasets. 
Merging both should fix this issue reliably.

**Comment by simonfuhrmann** [comment]
Merged #265, waiting for #270
